sentry-ruby/lib/sentry/metrics/aggregator.rb

```diff
@@ -0,0 +1,185 @@
+
+        # the flush interval needs to be shifted once per startup to create jittering
+        @flush_shift = Random.rand * ROLLUP_IN_SECONDS
+      end
+
+      def add(type,
+              key,
+              value,
+              unit: 'none',
+              tags: {},
+              timestamp: nil)
+        return unless ensure_thread
+        return unless METRIC_TYPES.keys.include?(type)
+
+        timestamp = timestamp.to_i if timestamp.is_a?(Time)
+        timestamp ||= Sentry.utc_now.to_i
+
+        # this is integer division and thus takes the floor of the division
+        # and buckets into 10 second intervals
+        bucket_timestamp = (timestamp / ROLLUP_IN_SECONDS) * ROLLUP_IN_SECONDS
+
+        serialized_tags = serialize_tags(get_updated_tags(tags))
+        bucket_key = [type, key, unit, serialized_tags]
+
+        @mutex.synchronize do
+          @buckets[bucket_timestamp] ||= {}
+
+          if @buckets[bucket_timestamp][bucket_key]
+            @buckets[bucket_timestamp][bucket_key].add(value)
+          else
+            @buckets[bucket_timestamp][bucket_key] = METRIC_TYPES[type].new(value)
+          end
+        end
+      end
+
+      def flush(force: false)
+        flushable_buckets = get_flushable_buckets!(force)
+        return if flushable_buckets.empty?
+
+        payload = serialize_buckets(flushable_buckets)
+        envelope = Envelope.new
+        envelope.add_item(
+          { type: 'statsd', length: payload.bytesize },
+          payload,
+          is_json: false
+        )
+
+        Sentry.background_worker.perform do
+          @client.transport.send_envelope(envelope)
+        end
+      end
+
+      def kill
+        log_debug('[Metrics::Aggregator] killing thread')
+
+        @exited = true
+        @thread&.kill
+      end
+
+      private
+
+      def ensure_thread
+        return false if @exited
+        return true if @thread&.alive?
+
+        @thread = Thread.new do
+          loop do
+            # TODO-neel-metrics use event for force flush later
+            sleep(FLUSH_INTERVAL)
+            flush
+          end
+        end
+
+        true
+      rescue ThreadError
+        log_debug('[Metrics::Aggregator] thread creation failed')
+        @exited = true
+        false
+      end
+
+      # important to sort for key consistency
+      def serialize_tags(tags)
+        tags.flat_map do |k, v|
+          if v.is_a?(Array)
+            v.map { |x| [k.to_s, x.to_s] }
+          else
+            [[k.to_s, v.to_s]]
+          end
+        end.sort
+      end
+
+      def get_flushable_buckets!(force)
+        @mutex.synchronize do
+          flushable_buckets = {}
+
+          if force
+            flushable_buckets = @buckets
+            @buckets = {}
+          else
+            cutoff = Sentry.utc_now.to_i - ROLLUP_IN_SECONDS - @flush_shift
+            flushable_buckets = @buckets.select { |k, _| k <= cutoff }
+            @buckets.reject! { |k, _| k <= cutoff }
+          end
+
```

Comment:
Calculating the new total weight of remaining buckets. I guess this is also coming in a later PR?

Reply:
yes